### PR TITLE
Reset `DedupeIntegration`'s `last-seen` if `before_send` dropped the event

### PR DESCRIPTION
**Pull request by sentrivana**
Imagine an app throws an exception twice, from different places. The first exception is dropped in the user's `before_send`. The second exception is not. Should the second exception appear in Sentry?

The current state is that it won't, since `DedupeIntegration` will take the first, dropped exception into account. When encountering the second exception, it'll consider it a duplicate and will drop it, even though the first exception never made it to Sentry.

In this PR, we reset `DedupeIntegration`'s `last-seen` if an event has been dropped by `before_send`, ensuring that the next exception will be reported.

Closes https://github.com/getsentry/sentry-python/issues/371

### CODE REVIEW

File: tests/test_basics.py [hunk 1]
```diff
@@ -710,6 +710,37 @@ def test_dedupe_event_processor_drop_records_client_report(
...
+    sentry_init(before_send=before_send)
+    events = capture_events()
+
+    exc = ValueError("aha!")
+    for _ in range(2):
+        # The first ValueError will be dropped by before_send. The second
+        # ValueError will be accepted by before_send, and should be sent to
+        # Sentry.
+        try:
+            raise exc
```
Comment (antonpirker) [inline tests/test_basics.py:737]: What happens if another unrelated exception is raised and captured between the first and second capture of `exc`?

**Review by antonpirker — COMMENTED** [review]
I guess there code be an edge case that causes problem. (see my comment)

File: tests/test_basics.py (hunk 1 above)
Comment (sentrivana) [reply tests/test_basics.py:737]: 1. A is dropped in `before_send`, dedupe sets `last-seen` to `None`
2. B is raised, dedupe lets it through since `last-seen` is `None`, and it sets B as `last-seen`
3. A is raised, dedupe lets it through since `last-seen` is set to B and A != B

The outside behavior (i.e., what is sent and what is dropped by dedupe) is the same as now in this case

Comment (sentrivana) [reply tests/test_basics.py:737]: Maybe I misunderstood the scenario though -- lemme know if you see a problem with this @antonpirker

Comment (antonpirker) [reply tests/test_basics.py:737]: You are right. I also tested it locally and my assumption is wrong. So everything ok with that PR!

**Review by antonpirker — APPROVED** [review]
looks good
